### PR TITLE
Updates for v2

### PR DESCRIPTION
**Pull request by yorhodes**
Using https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/1288 of monorepo

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -40,6 +40,12 @@ yarn
 yarn build
 ```
 
+### Using local core contracts
+```
+# yarn link does not work well with workspaces
+ln -s ~/path/to/mono/solidity node_modules/@hyperlane-xyz/core
```
Comment (jmrossy) [inline README.md:46]: Wasn't immediately obvious to me `mono` meant `monorepo`

File: README.md [hunk 2]
```diff
@@ -40,6 +40,12 @@ yarn
 yarn build
 ```
 
+### Using local core contracts
```
Comment (jmrossy) [inline README.md:43]: Maybe include a word/line that this is optional? It being in setup section makes it seem mandatory

File: contracts/HypERC20.sol [hunk 3]
```diff
@@ -10,96 +10,46 @@ import {ERC20Upgradeable} from "@openzeppelin/contracts-upgradeable/token/ERC20/
...
-        uint32 indexed origin,
-        address indexed recipient,
-        uint256 amount
-    );
-
+contract HypERC20 is ERC20Upgradeable, TokenRouter {
     /**
      * @notice Initializes the Hyperlane router, ERC20 metadata, and mints initial supply to deployer.
-     * @param _abacusConnectionManager The address of the connection manager contract.
+     * @param _mailbox The address of the mailbox contract.
```
Comment (jmrossy) [inline contracts/HypERC20.sol:16]: Happy to see this got renamed :)

File: contracts/HypERC20Collateral.sol [hunk 4]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.13;
+
+import {TokenRouter} from "./libs/TokenRouter.sol";
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/**
+ * @title Collateralize ERC20 token and route messages to HypERC20 tokens.
```
Comment (jmrossy) [inline contracts/HypERC20Collateral.sol:9]: Should this be @notice?

File: contracts/HypERC721Collateral.sol [hunk 5]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.13;
+
+import {TokenRouter} from "./libs/TokenRouter.sol";
+
+import {IERC721} from "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+
+/**
+ * @title Collateralize ERC20 token and route messages to HypERC20 tokens.
```
Comment (jmrossy) [inline contracts/HypERC721Collateral.sol:9]: ditto

File: contracts/libs/Message.sol [hunk 6]
```diff
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.8.0;
```
Comment (jmrossy) [inline contracts/libs/Message.sol:2]: intentionally different than files above?

File: contracts/libs/TokenRouter.sol [hunk 7]
```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.13;
+
+import {Router} from "@hyperlane-xyz/core/contracts/Router.sol";
+import {TypeCasts} from "@hyperlane-xyz/core/contracts/libs/TypeCasts.sol";
+import {Message} from "./Message.sol";
+
+/**
+ * @title Hyperlane Token that extends the ERC20 token standard to enable native interchain transfers.
```
Comment (jmrossy) [inline contracts/libs/TokenRouter.sol:9]: ditto

File: src/config.ts [hunk 8]
```diff
@@ -2,18 +2,21 @@ import { ethers } from 'ethers';
...
   name: string;
   symbol: string;
   totalSupply: ethers.BigNumberish;
 };
+export type CollateralConfig = {
+  token: string;
+}
 
-export type HypERC20Config = RouterConfig & Erc20TokenConfig;
+export const isCollateralConfig = (config: RouterConfig & TokenConfig | CollateralConfig): config is RouterConfig & CollateralConfig => {
```
Comment (jmrossy) [inline src/config.ts:14]: Is TS treating the config param type as `RouterConfig & (TokenConfig | CollateralConfig)` or `(RouterConfig & TokenConfig) | CollateralConfig` ?
My guess would be latter but code suggest former

Comment (jmrossy) [reply src/config.ts:14]: Also this approach is brittle (breaks if TokenConfig gets a token prop). Probably okay for now but some `_type` prop on which is do discriminated union would be more robust

File: src/deploy.ts [hunk 9]
```diff
@@ -1,79 +1,118 @@
...
 export class HypERC20Deployer<
-  Chain extends ChainName,
+  Chain extends ChainName // inferred from configured chains passed to constructor
 > extends HyperlaneRouterDeployer<
   Chain,
-  HypERC20Config,
+  HypERC20Config | HypERC20CollateralConfig,
   HypERC20Contracts,
-  HypERC20Factories
+  any
```
Comment (jmrossy) [inline src/deploy.ts:27]: intentional?

**Review by jmrossy — COMMENTED** [review]
Nice work! Mostly looks good, just some small stuff

File: contracts/libs/TokenRouter.sol (hunk 7 above)
Comment (yorhodes) [reply contracts/libs/TokenRouter.sol:9]: https://docs.soliditylang.org/en/latest/natspec-format.html#tags

**Review by jmrossy — APPROVED** [review]

**Pull request merged**